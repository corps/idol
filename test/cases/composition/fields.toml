[BasicStructOne]
fields.a = "enums.EnumOneTwo"
fields.b = "primitive_widens.IntAlias"

[BasicStructCoSelf]
is_a = ["BasicStruct", "BasicStruct"]

[BasicStructContraSelf]
is_a = ["BasicStruct", "BasicStruct"]

[BasicStructTwo]
fields.a = "enums.EnumOneTwo"
fields.c = "int"

[BasicStructOneTwoCo]
is_a = ["BasicStructOne", "BasicStructTwo"]
fields.d = "string[]"

[BasicStructOneTwoContra]
is_a = ["BasicStructOne", "BasicStructTwo"]
fields.d = "string[]"
variance = "Contravariant"

[BasicStructOneTwoContraTrim]
is_a = ["BasicStructOne", "BasicStructTwo"]
fields.d = "string[]"
variance = "Contravariant"
trim = true
