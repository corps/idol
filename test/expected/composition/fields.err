Processing fields...
Processing primitive_widens...
Processing enums...
Problem processing module fields: declaration BasicStructOneTwoCo: composition error: field b on left hand side has different variance than field c on the right.  Fields are not compatible
Error: 1
