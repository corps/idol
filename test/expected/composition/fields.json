{
  "BasicStructCoSelf": {
    "enum": [],
    "fields": null,
    "is_a": [
      "fields.BasicStructOne"
    ],
    "tags": [],
    "trim": false,
    "variance": "Covariant"
  },
  "BasicStructContraSelf": {
    "enum": [],
    "fields": null,
    "is_a": [
      "fields.BasicStructOne"
    ],
    "tags": [],
    "trim": false,
    "variance": "Covariant"
  },
  "BasicStructOne": {
    "enum": [],
    "fields": {
      "a": [
        "enums.EnumOneTwo"
      ],
      "b": [
        "primitive_widens.IntAlias"
      ]
    },
    "is_a": [],
    "tags": [],
    "trim": false,
    "variance": "Covariant"
  },
  "BasicStructOneTwoCo": {
    "enum": [],
    "fields": {
      "a": [
        "enums.EnumOneTwo"
      ],
      "b": [
        "primitive_widens.IntAlias"
      ],
      "c": [
        "int"
      ],
      "d": [
        "string[]"
      ]
    },
    "is_a": [],
    "tags": [],
    "trim": false,
    "variance": "Covariant"
  },
  "BasicStructOneTwoCoTrim": {
    "enum": [],
    "fields": {
      "b": [
        "primitive_widens.IntAlias"
      ],
      "c": [
        "int"
      ]
    },
    "is_a": [],
    "tags": [],
    "trim": false,
    "variance": "Covariant"
  },
  "BasicStructOneTwoContra": {
    "enum": [],
    "fields": {
      "a": [
        "enums.EnumOneTwo"
      ],
      "b": [
        "primitive_widens.IntAlias",
        "optional"
      ],
      "c": [
        "int",
        "optional"
      ]
    },
    "is_a": [],
    "tags": [],
    "trim": false,
    "variance": "Covariant"
  },
  "BasicStructOneTwoContraWithD": {
    "enum": [],
    "fields": {
      "a": [
        "enums.EnumOneTwo",
        "optional"
      ],
      "b": [
        "primitive_widens.IntAlias",
        "optional"
      ],
      "c": [
        "int",
        "optional"
      ],
      "d": [
        "string[]",
        "optional"
      ]
    },
    "is_a": [],
    "tags": [],
    "trim": false,
    "variance": "Covariant"
  },
  "BasicStructTwo": {
    "enum": [],
    "fields": {
      "a": [
        "enums.EnumOneTwo"
      ],
      "c": [
        "int"
      ]
    },
    "is_a": [],
    "tags": [],
    "trim": false,
    "variance": "Covariant"
  }
}
